trio/_socket.py

```diff
@@ -510,26 +511,213 @@ async def _resolve_address_nocp(
     if family == _stdlib_socket.AF_INET6:
         list_normed = list(normed)
         assert len(normed) == 4
-        # typechecking certainly doesn't like this logic, but given just how broad
-        # Address is, it's quite cumbersome to write the below without type: ignore
         if len(address) >= 3:
-            list_normed[2] = address[2]  # type: ignore
+            list_normed[2] = address[2]
         if len(address) >= 4:
-            list_normed[3] = address[3]  # type: ignore
-        return tuple(list_normed)  # type: ignore
+            list_normed[3] = address[3]
+        return tuple(list_normed)
     return normed
 
 
-# TODO: stopping users from initializing this type should be done in a different way,
-# so SocketType can be used as a type. Note that this is *far* from trivial without
-# breaking subclasses of SocketType. Can maybe add abstract methods to SocketType,
-# or rename _SocketType.
 class SocketType:
-    def __init__(self) -> NoReturn:
-        raise TypeError(
-            "SocketType is an abstract class; use trio.socket.socket if you "
-            "want to construct a socket object"
-        )
+    def __init__(self) -> None:
+        if type(self) == SocketType:
+            raise TypeError(
+                "SocketType is an abstract class; use trio.socket.socket if you "
+                "want to construct a socket object"
+            )
```

Comment:
Is there a reason you switched this to behind this `if`?

I think this makes sense but a bit confused what led to this change in specific.

Reply:
So this code doesn't error and require you to manually specify that you want to call `MyOtherType.__init__`:
```python
class MyOtherType:
  def __init__(self):
    """does something"""

class MySocketType(SocketType, MyOtherType):
  def __init__(self):
    super().__init__()
    """more code"""
```

Reply:
Oh that's interesting, I wasn't aware you could do that. What order does it call both init functions in? MRO?

Reply:
The definition here should probably also call `super()`, so it chains to `MyOtherType` too.

Reply:
> Oh that's interesting, I wasn't aware you could do that. What order does it call both init functions in? MRO?

MRO, yes https://stackoverflow.com/questions/3277367/how-does-pythons-super-work-with-multiple-inheritance